source/plugin/data/keys.rst

```diff
@@ -0,0 +1,107 @@
+
+    public void heal(DataHolder target) {
+        if (target.supports(Keys.HEALTH)) {
+            double maxHealth = target.get(Keys.MAX_HEALTH).get();
+            target.offer(Keys.HEALTH, maxHealth);
+        }
+    }
+
+Now for the details of the above function.
+
+The first line checks if our given data holder supports a current health value. Only if it does, it can be healed after
+all. Since a data holder can not have current health without having a maximum health and vice versa, a check for
+one of the keys using the ``supports()`` method suffices.
+
+The second line uses the ``getOrNull()`` function to ask the data holder for its maximum health. Besides
+``getOrNull()``, the methods ``get()`` and ``getOrElse()`` exist, all of which accept a ``Key`` as their first
+parameter. Generally, ``get()`` should be used, which will return an ``Optional`` of the data requested or
+``Optional.absent()`` if the data holder does not support the supplied key. Since we already verified that the
+``Key`` is supported, we can just call ``get()`` on the Optional without further checks. We could also use
+``getOrNull()`` which is basically a shortcut to call ``get(key).orNull()``, thus getting rid of the
+``Optional``. The third possibility would be the ``getOrElse()``, which accepts a default value as a second
+parameter to be returned if the value is not present on the data holder.
+
+In the third line, we offer data back to the data holder. We provide a ``Key`` denoting the current health and the
+before acquired maximum health, thus healing the data holder to full health. There are a variety of ``offer()``
+methods accepting different parameter sets, all of which return a ``DataTransactionResult`` containing information
+if the offer was accepted. For now, we'll use the one accepting a ``Key`` and a corresponding value, but we will
+encounter more in the next pages. Since we already know that our offer of current health is accepted (as the data
+holder supports it), we can silently discard the result.
+
+Transforming Data
+=================
+
+Other than getting, modifying and offering a value, there is another way of interacting with data. Using a data
+holder's ``transform()`` method we can pass a ``Key`` and a ``Function``. Internally, the value for the key will be
+retrieved and the given function applied to it. The result is then stored under the key and the ``transform()``
+method will return a ``DataTransactionResult`` accordingly.
+
+Now, as an example, imagine we want to buff a data holder by doubling his maximum health.
+
+.. code-block:: java
+
+    public void buff(DataHolder target) {
+        target.transform(Keys.MAX_HEALTH, new Function<Double,Double>() {
+            @Override
+            public Double apply(Double input) {
+                return (input == null) ? 0 : input * 2;
+            }
+        });
+    }
+
+Or, since we use Java 8 and are able to make use of its lambda expressions:
```

Comment:
"Or if you use Java 8, you're able to shorten the line with lambda expressions:"